azure/scope/machine.go

```diff
@@ -289,17 +291,17 @@ func (m *MachineScope) DiskSpecs() []azure.ResourceSpecGetter {
 }
 
 // RoleAssignmentSpecs returns the role assignment specs.
-func (m *MachineScope) RoleAssignmentSpecs() []azure.RoleAssignmentSpec {
+func (m *MachineScope) RoleAssignmentSpecs() []azure.ResourceSpecGetter {
+	roles := make([]azure.ResourceSpecGetter, 1)
 	if m.AzureMachine.Spec.Identity == infrav1.VMIdentitySystemAssigned {
-		return []azure.RoleAssignmentSpec{
-			{
-				MachineName:  m.Name(),
-				Name:         m.AzureMachine.Spec.RoleAssignmentName,
-				ResourceType: azure.VirtualMachine,
-			},
+		roles[0] = &roleassignments.RoleAssignmentSpec{
```

Comment:
This would be a good spot to assign these values so we don't need to in `Reconcile()`. Also that would mean you don't need to mock the `s.Scope.SubscriptionID()` return values in the test cases.

```
scope := fmt.Sprintf("/subscriptions/%s/", s.Scope.SubscriptionID())
rs.Scope = scope
contributorRoleDefinitionID := fmt.Sprintf("/subscriptions/%s/providers/Microsoft.Authorization/roleDefinitions/%s", s.Scope.SubscriptionID(), azureBuiltInContributorID)
rs.RoleDefinitionID = contributorRoleDefinitionID
```

Reply:
Added the values here. Only downside is in machine scope unit testing. 
See this https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1873/files#r794443824